core/vm/logger.go

```diff
@@ -29,15 +30,18 @@ import (
 // if you need to retain them beyond the current call.
 type EVMLogger interface {
 	// Transaction level
-	CaptureTxStart(gasLimit uint64)
-	CaptureTxEnd(restGas uint64)
+	CaptureTxStart(evm *EVM, tx *types.Transaction)
+	CaptureTxEnd(receipt *types.Receipt, err error)
 	// Top call frame
-	CaptureStart(env *EVM, from common.Address, to common.Address, create bool, input []byte, gas uint64, value *big.Int)
+	CaptureStart(from common.Address, to common.Address, create bool, input []byte, gas uint64, value *big.Int)
 	CaptureEnd(output []byte, gasUsed uint64, err error)
 	// Rest of call frames
 	CaptureEnter(typ OpCode, from common.Address, to common.Address, input []byte, gas uint64, value *big.Int)
 	CaptureExit(output []byte, gasUsed uint64, err error)
 	// Opcode level
 	CaptureState(pc uint64, op OpCode, gas, cost uint64, scope *ScopeContext, rData []byte, depth int, err error)
 	CaptureFault(pc uint64, op OpCode, gas, cost uint64, scope *ScopeContext, depth int, err error)
+	CaptureKeccakPreimage(hash common.Hash, data []byte)
+	// Misc
+	OnGasConsumed(gas, amount uint64)
```

Comment:
I agree that gas refund needs to be tracked to get an accurate view of gas consumption. 

I don't personally see the need to have a change on the initial "gas limit", but I can see that it would closes the full loop of gas tracking, so that you start with 0 and ends with 0.

Reply:
I would be against `SetBalanceWithReason` as we could forget to add reason in new places. Actually `0x0` is mostly used in tests. I think you found the only place outside of tests that we pass `0` to `SetBalance` 😁

I can update them to use the `state.BalanceChangeUnspecified` reason to make it more clear what that parameter is.